src/tree_annotation/database.cljs

```diff
@@ -25,172 +27,323 @@
+
+(defn rename-node [node label]
+  (assoc node :label label))
+
+(defn rename-node! [label index]
+  "Assings the label `label` to the node at `index`."
+  (swap! db update :forest update-forest #(rename-node % label) index))
+
+(defn start-renaming [node]
+  (assoc node :renaming true))
+
+(defn start-renaming-node! [index]
+  "Puts the node at `index` into renaming mode."
+  (swap! db update :forest update-forest start-renaming index))
+
+(defn stop-renaming [node]
+  (assoc node :renaming false))
+
+(defn stop-renaming-node! [index]
+  "Puts the node at `index` out of renaming mode."
+  (swap! db update :forest update-forest stop-renaming index))
+
+;; combine
+;; -------
+
+(defn combine [children]
+  "Combines the sequence `children` into a new tree."
+  (let [label (:label (last children))]
+    (assoc default-node
+           :label label
+           :children (mapv deselect-tree children)
+           :x (:x (first children))
+           :y (inc (reduce max (map :y children)))
+           :width (reduce + (map :width children)))))
+
+(defn combine-selected
+  ([forest]
+   "Finds groups of adjacent selected trees in `forest` and combines them."
+   (combine-selected forest [] []))
+  ([forest group acc]
+   (letfn [(combine-if [coll xs]
+             (if (empty? xs) coll (conj coll (combine xs))))]
+     (if (empty? forest)
+       ; end of forest
+       (combine-if acc group)
+       ; in forest
+       (let [tree (first forest)
+             tail (rest forest)]
+         (if (tree-selected? tree)
+           (combine-selected tail (conj group tree) acc)
+           (combine-selected tail [] (conj (combine-if acc group) tree))))))))
+
+(defn combine-selected! []
+  "Finds groups of adjacent selected trees and combines them."
+  (swap! db update :forest combine-selected))
+
+;; delete
+;; ------
+
+(defn node-delete-selected [node]
+  "Deletes all selected children of `node`.
+Returns either the unchanged node or a list of remaining subtrees."
+  (let [subtrees (mapv node-delete-selected (:children node))]
+    (if (or (and (:selected node) (not (leaf? node)))
+            (not-every? map? subtrees))
+      (vec (flatten subtrees))
+      node)))
+
+(defn delete-selected [forest]
+  "Delete all selected nodes and their ancestors from `forest`."
+  (vec (flatten (map node-delete-selected forest))))
+
+(defn delete-selected! []
+  "Delete all selected nodes and their ancestors."
+  (swap! db update :forest delete-selected))
+
+;; parse
+;; -----
+
+(declare recalc-coords)
+
+(defn recalc-coords-tree [node offset]
+  "Recalculate the coordinates in a tree with x offset `offset`."
+  (let [[children' offset-children] (recalc-coords (:children node) offset)
+        offset' (if (leaf? node) (inc offset-children) offset-children)
+        node' (assoc node
+                     :children children'
+                     :x offset
+                     :y (inc (reduce max 0 (map :y children')))
```

Comment:
Not sure what you mean. If we know the y coordinate of all children, the y coordinate of the parent must be (maximum y of its children) + 1. We cannot take the y of the input node, since we are recalculating the coordinates, so we can't rely on any old values.